addons/io_scene_gltf2/blender/exp/gltf2_blender_image.py

```diff
@@ -14,121 +14,198 @@
+                    result[:, :, dst_chan] = pixels[:, :, img_fill.src_chan]
+
+            pixels = None  # GC this please
+
+        if result is None:
+            # No ImageFills; use a 1x1 white pixel
+            result = np.array([1.0, 1.0, 1.0, 1.0])
+            result = result.reshape((1, 1, 4))
+
+        return self.__encode_from_numpy_array(result)
+
+    def __encode_from_numpy_array(self, array: np.ndarray) -> bytes:
+        tmp_image = None
+        try:
+            tmp_image = bpy.data.images.new(
+                "##gltf-export:tmp-image##",
+                width=array.shape[0],
+                height=array.shape[1],
+                alpha=Channel.A in self.fills,
+            )
+            assert tmp_image.channels == 4  # 4 regardless of the alpha argument above.
+
+            # Also slow and eats all your memory.
+            tmp_image.pixels = array.flatten().tolist()
+
+            return _encode_temp_image(tmp_image, self.file_format)
+
+        finally:
+            if tmp_image is not None:
+                bpy.data.images.remove(tmp_image, do_unlink=True)
+
+    def __encode_from_image(self, image: bpy.types.Image) -> bytes:
+        # See if there is an existing file we can use.
+        if self.file_format == image.file_format:
+            src_path = bpy.path.abspath(image.filepath_raw)
             if os.path.isfile(src_path):
-                with open(src_path, "rb") as f:
-                    encoded_image = f.read()
-                return encoded_image
+                with open(src_path, 'rb') as f:
+                    return f.read()
+
+        # Copy to a temp image and save.
+        tmp_image = None
+        try:
+            tmp_image = image.copy()
+            return _encode_temp_image(tmp_image, self.file_format)
+        finally:
+            if tmp_image is not None:
+                bpy.data.images.remove(tmp_image, do_unlink=True)
 
-        image = bpy.data.images.new("TmpImage", width=self.width, height=self.height, alpha=self._has_alpha)
-        pixels = self._img.flatten().tolist()
-        image.pixels = pixels
 
-        # we just use blenders built in save mechanism, this can be considered slightly dodgy but currently is the only
-        # way to support
-        with tempfile.TemporaryDirectory() as tmpdirname:
-            tmpfilename = tmpdirname + "/img"
-            image.filepath_raw = tmpfilename
-            image.file_format = file_format
-            image.save()
+def _encode_temp_image(tmp_image: bpy.types.Image, file_format: str) -> bytes:
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        tmpfilename = tmpdirname + '/img'
+        tmp_image.filepath_raw = tmpfilename
 
-            with open(tmpfilename, "rb") as f:
-                encoded_image = f.read()
+        # NOT A TYPO!!! If you delete this line, the
+        # assignment on the next line will not work.
+        tmp_image.file_format
+        tmp_image.file_format = file_format
```

Comment:
Believe it or not.... :)

![insanity](https://user-images.githubusercontent.com/11024420/70366857-4a422f80-1860-11ea-9176-377292bf4abd.png)

I can reproduce with both today's 2.82 alpha and 2.79.

Other than that, this is the same way master saves an image.